api/scpca_portal/models/sample.py

```diff
@@ -265,6 +265,71 @@ def single_cell_file_formats(self):
             file_formats.append(ComputedFile.OutputFileFormats.ANN_DATA)
         return file_formats
 
+    def get_computed_file_name_from_download_config(self, download_config: Dict):
+        match download_config:
+            case {"modality": "SPATIAL"}:
+                return self.output_spatial_computed_file_name
+            case {"format": "ANN_DATA"}:
+                return self.output_single_cell_anndata_computed_file_name
+            case {"modality": "SINGLE_CELL"}:
+                return (
+                    self.output_single_cell_computed_file_name
+                    if not self.has_multiplexed_data
+                    else self.output_multiplexed_computed_file_name
+                )
+
+    def create_computed_files(
+        self,
+        project,
+        max_workers=8,  # 8 = 2 file formats * 4 mappings.
+        clean_up_output_data=True,
+        update_s3=False,
+    ):
+        """Prepares ready for saving project computed files based on generated file mappings."""
+
+        def on_get_sample_file(future):
+            computed_file = future.result()
+            if computed_file:
+                computed_file.process_computed_file(clean_up_output_data, update_s3)
+
+        samples = Sample.objects.filter(project__scpca_id=project.scpca_id)
+        logger.info(
+            f"Processing {len(samples)} sample{pluralize(len(samples))} using "
+            f"{max_workers} worker{pluralize(max_workers)}"
+        )
+
+        # Prepare a threading.Lock for each sample, with the chief purpose being to protect
+        # multiplexed samples that shares a zip file.
+        # The keys are the sample.multiplexed_ids since that will be unique across shared zip files.
+        non_multiplexed_ids = set(s.scpca_id for s in samples)
+        multiplexed_ids = set(
+            ["_".join(s.multiplexed_ids) for s in samples if s.has_multiplexed_data]
+        )
+        locks = {
+            f'{id_entry}-{config["modality"]}-{config["format"]}': Lock()
+            for id_entry in non_multiplexed_ids.union(multiplexed_ids)
+            for config in common.GENERATED_SAMPLE_DOWNLOAD_CONFIGURATIONS
+        }
```

Comment:
Maybe, instead of all this add a property?

```python
@property
def lock_prefix(self):
    if self.is_multiplexed:
       return "_".join(self.multiplexed_ids)
    return self.scpca_id
```

Then:

```suggestion
        lock_prefixes = set([s.lock_prefix for s in samples])
        locks = {
            f'{lock_prefix}-{config["modality"]}-{config["format"]}': Lock()
            for lock_prefix in lock_prefixes
            for config in common.GENERATED_SAMPLE_DOWNLOAD_CONFIGURATIONS
        }
```